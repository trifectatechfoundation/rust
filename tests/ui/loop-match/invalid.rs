#![feature(loop_match)]
#![crate_type = "lib"]

enum State {
    A,
    B,
    C,
}

fn invalid_update() {
    let mut fake = State::A;
    let state = State::A;
    #[loop_match]
    loop {
        fake = 'blk: {
            //~^ ERROR: invalid update of the `loop_match` state
            match state {
                _ => State::B,
            }
        }
    }
}

fn invalid_scrutinee() {
    let state = State::A;
    #[loop_match]
    loop {
        state = 'blk: {
            match State::A {
                //~^ ERROR: invalid match on `loop_match` state
                _ => State::B,
            }
        }
    }
}

fn bad_statements_1() {
    let state = State::A;
    #[loop_match]
    loop {
        1;
        //~^ ERROR: statements are not allowed in this position within a `loop_match`
        state = 'blk: {
            match State::A {
                _ => State::B,
            }
        }
    }
}

fn bad_statements_2() {
    let state = State::A;
    #[loop_match]
    loop {
        state = 'blk: {
            1;
            //~^ ERROR: statements are not allowed in this position within a `loop_match`
            match State::A {
                _ => State::B,
            }
        }
    }
}

fn bad_rhs_1() {
    let state = State::A;
    #[loop_match]
    loop {
        state = State::B
        //~^ ERROR: this expression must be a single `match` wrapped in a labelled block
    }
}

fn bad_rhs_2() {
    let state = State::A;
    #[loop_match]
    loop {
        state = 'blk: {
            State::B
            //~^ ERROR: this expression must be a single `match` wrapped in a labelled block
        }
    }
}

fn bad_rhs_3() {
    let state = ();
    #[loop_match]
    loop {
        state = 'blk: {
            //~^ ERROR: this expression must be a single `match` wrapped in a labelled block
        }
    }
}

fn missing_assignment() {
    let state = State::A;
    #[loop_match]
    loop {
        () //~ ERROR:  expected a single assignment expression
    }
}

fn empty_loop_body() {
    let state = State::A;
    #[loop_match]
    loop {
        //~^ ERROR:  expected a single assignment expression
    }
}

fn break_without_value() {
    let state = State::A;
    #[loop_match]
    'a: loop {
        state = 'blk: {
            match state {
                State::A => {
                    #[const_continue]
                    break 'blk;
                    //~^ ERROR: mismatched types
                }
                _ => break 'a,
            }
        }
    }
}

fn break_without_value_unit() {
    let state = ();
    #[loop_match]
    'a: loop {
        state = 'blk: {
            match state {
                () => {
                    #[const_continue]
                    break 'blk;
                    //~^ ERROR: a `const_continue` must break to a label with a value
                }
            }
        }
    }
}
