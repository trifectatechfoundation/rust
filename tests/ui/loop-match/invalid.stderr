error[E0308]: mismatched types
  --> $DIR/invalid.rs:121:21
   |
LL |                     break 'blk;
   |                     ^^^^^^^^^^ expected `State`, found `()`
   |
help: give the `break` a value of the expected type
   |
LL |                     break 'blk /* value */;
   |                                +++++++++++

error: invalid update of the `loop_match` state
  --> $DIR/invalid.rs:17:9
   |
LL |         fake = 'blk: {
   |         ^^^^
LL |
LL |             match state {
   |                   ----- the assignment must update this variable

error: invalid match on `loop_match` state
  --> $DIR/invalid.rs:31:19
   |
LL |             match State::A {
   |                   ^^^^^^^^
   |
   = note: only matches on local variables are valid

error: statements are not allowed in this position within a `#[loop_match]`
  --> $DIR/invalid.rs:43:9
   |
LL |         1;
   |         ^^

error: statements are not allowed in this position within a `#[loop_match]`
  --> $DIR/invalid.rs:58:13
   |
LL |             1;
   |             ^^

error: this expression must be a single `match` wrapped in a labeled block
  --> $DIR/invalid.rs:71:17
   |
LL |         state = State::B
   |                 ^^^^^^^^

error: this expression must be a single `match` wrapped in a labeled block
  --> $DIR/invalid.rs:81:13
   |
LL |             State::B
   |             ^^^^^^^^

error: this expression must be a single `match` wrapped in a labeled block
  --> $DIR/invalid.rs:91:17
   |
LL |           state = 'blk: {
   |  _________________^
LL | |
LL | |         }
   | |_________^

error: expected a single assignment expression
  --> $DIR/invalid.rs:101:9
   |
LL |         ()
   |         ^^

error: expected a single assignment expression
  --> $DIR/invalid.rs:108:10
   |
LL |       loop {
   |  __________^
LL | |
LL | |     }
   | |_____^

error: a `#[const_continue]` must break to a label with a value
  --> $DIR/invalid.rs:138:21
   |
LL |                     break 'blk;
   |                     ^^^^^^^^^^

error: match arms that are part of a `#[loop_match]` cannot have guards
  --> $DIR/invalid.rs:156:29
   |
LL |                 State::B if cond => break 'a,
   |                             ^^^^

error: aborting due to 12 previous errors

For more information about this error, try `rustc --explain E0308`.
